testing/trino-testing-services/src/main/java/io/trino/testng/services/ReportBadJunitTestAnnotations.java

```diff
@@ -0,0 +1,125 @@
+ * limitations under the License.
+ */
+package io.trino.testng.services;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestInstanceFactoryContext;
+import org.junit.jupiter.api.extension.TestInstancePreConstructCallback;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Throwables.getStackTraceAsString;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.testng.services.Listeners.reportListenerFailure;
+import static java.util.stream.Collectors.joining;
+
+public class ReportBadJunitTestAnnotations
+        implements TestInstancePreConstructCallback
+{
+    @Override
+    public void preConstructTestInstance(TestInstanceFactoryContext factoryContext, ExtensionContext context)
+    {
+        Class<?> testClass = factoryContext.getTestClass();
+        try {
+            reportBadTestAnnotations(testClass);
+        }
+        catch (RuntimeException | Error e) {
+            reportListenerFailure(
+                    ReportBadTestAnnotations.class,
+                    "Failed to process %s: \n%s",
+                    testClass,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    private void reportBadTestAnnotations(Class<?> testClass)
+    {
+        List<Method> unannotatedTestMethods = findUnannotatedInheritedTestMethods(testClass);
+        if (!unannotatedTestMethods.isEmpty()) {
+            reportListenerFailure(
+                    ReportBadJunitTestAnnotations.class,
+                    "Test class %s has methods which are inherited but not explicitly annotated. Are they missing @Test?%s",
+                    testClass.getName(),
+                    unannotatedTestMethods.stream()
+                            .map(Method::toString)
+                            .collect(joining("\n\t\t", "\n\t\t", "")));
+        }
+    }
+
+    @VisibleForTesting
+    static List<Method> findUnannotatedInheritedTestMethods(Class<?> realClass)
+    {
+        return Arrays.stream(realClass.getMethods())
+                .filter(method -> method.getDeclaringClass() != Object.class)
+                .filter(method -> !Modifier.isStatic(method.getModifiers()))
+                .filter(method -> !method.isBridge())
+                .filter(method -> !isUnannotated(method) && overriddenMethodHasTestAnnotation(method))
+                .collect(toImmutableList());
+    }
+
+    private static boolean isUnannotated(Method method)
+    {
+        return Arrays.stream(method.getAnnotations()).map(Annotation::annotationType)
+                .anyMatch(ReportBadJunitTestAnnotations::isJUnitAnnotation);
```

Comment:
isUnannotated = ... anyMatch(isJUnitAnnotation)

was `noneMatch` meant? or was the method be named `isAnnotated`?